component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/utils/TokenExchangeUtils.java

```diff
@@ -0,0 +1,503 @@
+                    + jwtIssuer);
+        }
+        return identityProvider;
+    }
+
+    public static void handleException(String code, String errorMessage) throws IdentityOAuth2Exception {
+        log.error(errorMessage);
+        throw new IdentityOAuth2Exception(code, errorMessage);
+    }
+
+    public static void handleException(String errorMessage) throws IdentityOAuth2Exception {
+        log.error(errorMessage);
+        throw new IdentityOAuth2Exception(errorMessage);
+    }
+
+    /**
+     * Get token endpoint alias
+     *
+     * @param identityProvider Identity provider
+     * @return token endpoint alias
+     */
+    public static String getTokenEndpointAlias(IdentityProvider identityProvider, String tenantDomain) {
+
+        Property oauthTokenURL = null;
+        String tokenEndPointAlias = null;
+        if (IdentityApplicationConstants.RESIDENT_IDP_RESERVED_NAME.equals(
+                identityProvider.getIdentityProviderName())) {
+            try {
+                identityProvider = IdentityProviderManager.getInstance().getResidentIdP(tenantDomain);
+            } catch (IdentityProviderManagementException e) {
+                log.debug("Error while getting Resident IDP :" + e.getMessage());
+            }
+            FederatedAuthenticatorConfig[] fedAuthnConfigs =
+                    identityProvider.getFederatedAuthenticatorConfigs();
+            FederatedAuthenticatorConfig oauthAuthenticatorConfig =
+                    IdentityApplicationManagementUtil.getFederatedAuthenticator(fedAuthnConfigs,
+                            IdentityApplicationConstants.Authenticator.OIDC.NAME);
+
+            if (oauthAuthenticatorConfig != null) {
+                oauthTokenURL = IdentityApplicationManagementUtil.getProperty(
+                        oauthAuthenticatorConfig.getProperties(),
+                        IdentityApplicationConstants.Authenticator.OIDC.OAUTH2_TOKEN_URL);
+            }
+            if (oauthTokenURL != null) {
+                tokenEndPointAlias = oauthTokenURL.getValue();
+                log.debug("Token End Point Alias of Resident IDP :" + tokenEndPointAlias);
+            }
+        } else {
+            tokenEndPointAlias = identityProvider.getAlias();
+            log.debug("Token End Point Alias of the Federated IDP: " + tokenEndPointAlias);
+        }
+        return tokenEndPointAlias;
+    }
+
+    /**
+     * Method to validate the signature of the JWT
+     *
+     * @param signedJWT signed JWT whose signature is to be verified
+     * @param idp       Identity provider who issued the signed JWT
+     * @return whether signature is valid, true if valid else false
+     * @throws com.nimbusds.jose.JOSEException
+     * @throws org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception
+     */
+    public static boolean validateSignature(SignedJWT signedJWT, IdentityProvider idp, String tenantDomain) throws
+            JOSEException, IdentityOAuth2Exception {
```

Comment:
can we use this util method? https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/JWTSignatureValidationUtils.java#L67

Reply:
The initial implementation of token exchange is planned to be compatible with IS 5.11. The above mentioned utils class seems to be there in master. We decided to use the methods inside `TokenExchangeUtils` class (will result in code duplication for now) and this will be removed in the future when upgrading to 5.12 compatible dependencies.